katdal/applycal.py

```diff
@@ -0,0 +1,188 @@
+
+    Raises
+    ------
+    KeyError
+        If input and/or cal product has no associated correction
+    """
+    g_per_input = np.ones((len(inputs), len(channels)), dtype='complex64')
+    for product in cal_products:
+        for n, inp in enumerate(inputs):
+            sensor_name = 'Calibration/{}_correction_{}'.format(inp, product)
+            g_product = cache.get(sensor_name)[dump]
+            if np.shape(g_product) != ():
+                g_product = g_product[channels]
+            g_per_input[n] *= g_product
+    g_per_cp = g_per_input[input1_index] * g_per_input[input2_index].conj()
+    return g_per_cp.T
+
+
+def apply_vis_correction(out, correction):
+    """Apply `correction` in-place to visibility data in `out`."""
+    correction[np.isnan(correction)] = np.complex64(1)
+    out *= correction
+
+
+def add_applycal_transform(indexer, cache, corrprods, cal_products,
+                           apply_correction):
+    """Add transform to indexer that applies calibration corrections.
+
+    This adds a transform to the indexer which wraps the underlying data
+    (visibilities, weights or flags). The transform will apply all calibration
+    corrections specified in `cal_products` to each dask chunk individually.
+    The actual application method is also user-specified, which allows most
+    of the machinery to be reused between visibilities, weights and flags.
+    The time and frequency selections are salvaged from `indexer` but the
+    selected `corrprods` still needs to be passed in as a parameter to identify
+    the relevant inputs in order to access correction sensors.
+
+    Parameters
+    ----------
+    indexer : :class:`katdal.lazy_indexer.DaskLazyIndexer` object
+        Indexer with underlying dask array that will be transformed
+    cache : :class:`katdal.sensordata.SensorCache` object
+        Sensor cache, used to look up individual correction sensors
+    corrprods : sequence of (string, string)
+        Selected correlation products as pairs of correlator input labels
+    cal_products : sequence of string
+        Calibration products that will contribute to corrections
+    apply_correction : function, signature ``out = f(out, correction)``
+        Function that will actually apply correction to data from indexer
+    """
+    if indexer.dataset.chunks[2] != (len(corrprods),):
+        raise ValueError('Applycal currently assumes a single chunk '
+                         'along the corrprod axis')
+    chunk_starts = [np.cumsum((0,) + c) for c in indexer.dataset.chunks]
+    dump_index = indexer.keep[0].nonzero()[0]
+    channel_index = indexer.keep[1].nonzero()[0]
+    # Turn corrprods into a list of input labels and two lists of indices
+    inputs = sorted(set(np.ravel(corrprods)))
+    input1_index = [inputs.index(cp[0]) for cp in corrprods]
+    input2_index = [inputs.index(cp[1]) for cp in corrprods]
+    ccpc_args = (cache, inputs, input1_index, input2_index, cal_products)
```

Comment:
Should make a copy of `cal_products`, so that if the call modifies their copy it won't affect future usage of our copy.

Reply:
Did you mean "the caller"? Good point.

Reply:
Done.